Adds coverage for product deletion and cloning

filipefurtad0 opened the pull request:
#### What? Why?

Closes #6638.

<!-- Explain why this change is needed and the solution you propose.
Provide context for others to understand it. -->

Adds basic coverage to product deletion and cloning, as feature specs.
Checks if line items from deleted products are displayed correctly on shipped orders (admin view).

#### What should we test?
<!-- List which features should be tested and how. -->

Green build.

#### Release notes
<!-- Write a one liner description of the change to be included in the release notes.
Every PR is worth mentioning, because you did it for a reason. -->

Covers the missing CRUD operations on the feature level (deletion and cloning).

<!-- Please select one for your PR and delete the other. -->
Changelog Category: Technical changes

#### Documentation updates
<!-- Are there any wiki pages that need updating after merging this PR?
List them here or remove this section. -->

None.

sauloperez commented on spec/features/admin/products_spec.rb:
```diff
@@ -114,6 +114,82 @@
     end
   end
+  
+  describe "soft-deleting" do
```
I would keep this add the abstraction level of features so this doesn't get out of date if we no longer implement it with soft-deletion.

filipefurtad0 replied:
Removed the `soft-` part of it.

sauloperez commented on spec/features/admin/products_spec.rb:
```diff
@@ -114,6 +114,82 @@
+      end
+    end
+
+    context 'a shipped product', js: true do
```
if we really need JS capabilities (have you tried without `js: true`) we can as well move this option to the `describe` above and it'll apply to all.

filipefurtad0 replied:
It does need JS capabilities, but I've moved it to the describe block, as suggested. I'm not really sure when JS capabilities are needed... (further reading for me :nerd_face: )

sauloperez commented on spec/features/admin/products_spec.rb:
```diff
@@ -114,6 +114,82 @@
+        # bug #660
+        # only the first variant is duplicated
+        expect(Spree::Product.second.variants.count).to eq 1
+        # TODO expect(Spree::Product.second.variants.count).to eq Spree::Product.first.variants.count
```
IMO these inner details of how cloning works belong to unit-level testing. This test should ensure the feature works but all possible code paths and cases should be covered by unit tests. However, I bet there's no single class/method dealing with cloning and it may not be obvious how to implement them. I haven't checked.

filipefurtad0 replied:
Agree, that makes sense. Added a second commit, moving these tests from the feature spec to the controller spec.

andrewpbrett reviewed (approved):
🙏 thanks Filipe!

jibees reviewed (approved):
I think it just needs a little cleanup of git history.
👍

sauloperez commented on spec/features/admin/products_spec.rb:
```diff
@@ -114,6 +114,71 @@
+      before do
+        login_as_admin_and_visit spree.admin_products_path
+
+        within "tr#p_#{product2.id}" do
```
I know we're using it in all feature specs but it'd be less redundant and hence robust to use `#p_#{product.id}` as selector. `#` selects id attributes and these are meant to be unique in HTML. As a result, we rely a bit less on the actual markup which will make these tests less error prone.

This is something I'd like to enforce. I've had a good experiencie doing this in other projects.

sauloperez replied:
thoughts @andrewpbrett @Matt-Yorkley @jibees ?

andrewpbrett replied:
The difference is just that you're removing the `tr` in case this is one day no longer a table? That makes sense to me, yeah.

jibees replied:
I totally agree.

sauloperez replied:
exactly, feature specs don't care whether or not this is a table, just if it works.

filipefurtad0 replied:
Removed the `tr` tag :+1:

sauloperez commented on spec/features/admin/products_spec.rb:
```diff
@@ -114,6 +114,71 @@
+        expect(page).to have_selector "tr#p_#{product1.id}"
+        expect(page).not_to have_selector "tr#p_#{order.variants.first.product_id}"
+        expect(Spree::Product.count).to eq 1
+        expect(order.line_items.count).to eq 1
```
Likewise, I found it also leads to more robust feature specs to rely on UI only at this testing-level. The idea is that everything is checked counting UI elements, or navigating to other index pages that do show the information we need. That's what we do when testing manually and is less dependant on the implementation underneath. Therefore, it gives us more confidence while refactoring.

filipefurtad0 replied:
That's a good idea; I removed those two lines as the result might be misleading. As you say, the idea here is to test the UI.

I felt it was a good idea to add a `visit current_path` in the respective `before` block. I think the assertion has value only after page reload: this way we are sure that the DB contents are being used to populate the page.

sauloperez reviewed (commented):
I'm thinking we could use RSpec's `pending` and/or tag tests to better track those that are meant to be fixed.

Matt-Yorkley commented on spec/controllers/api/v0/products_controller_spec.rb:
```diff
@@ -183,6 +184,27 @@
+      it '(does not) clone price' do
+        product.update_attribute(:price, 2.22)
+        spree_post :clone, product_id: product.id, format: :json
+        expect(json_response['price']).not_to eq(2.22)
```
It doesn't clone the price? That seems really strange to me! What price does it actually have, in this case..?

filipefurtad0 replied:
It seems to clone the price from the master variant, for that product. So, if the price of product is changed, the initial price will be considered, and not the value after updating. See also [this comment](https://github.com/openfoodfoundation/openfoodnetwork/issues/660#issuecomment-137298238)

filipefurtad0 commented:
Ok I messed-up a commit (doh!) and reverted it; after that I cleaned up that mess from the branch history (seems to have worked).

> I think it just needs a little cleanup of git history.

But I'm not entirely sure how to clean the history regarding the other commits, as they seem (to me) to address the review... I'll DM you @jibees later on that, for advice :+1: Thanks!

filipefurtad0 commented:
Alright! Thanks for your help @jibees. I've squashed those minor commits - and left the first and last one. Please let me know if there is still something which I can improve here.

jibees commented:
> Alright! Thanks for your help @jibees. I've squashed those minor commits - and left the first and last one. Please let me know if there is still something which I can improve here.

@filipefurtad0 
You should rename your commit : https://github.com/openfoodfoundation/openfoodnetwork/pull/7402/commits/a794f94cee9ee0fa4776dbb6e1149f16ae1d3d64 as it contains information that are unnecessary: `squashes commits;`
You should find useful information on creating a great commit, and specially around message: 
https://github.com/openfoodfoundation/openfoodnetwork/wiki/Making-a-great-commit#commit-messages-should

jibees commented:
Next step should be merging? No need to go to `Dev test ready`?

filipefurtad0 commented:
Yes, it's just specs, so I guess no-test here. Moving to ready to go :+1: Thanks @jibees !
